### PR TITLE
Refactoring IShiftOperators to take a `TOther`

### PR DESCRIPTION
**Pull request by tannergooding**
This is failing locally, checking if CI provides any additional details.

### CODE REVIEW

**Comment by tannergooding** [comment]
@davidwrighton, @lambdageek seeing:
```
Assert failure(PID 27404 [0x00006b0c], Thread: 7640 [0x1dd8]): pMT != NULL

CORECLR! CoreLibBinder::GetExistingClass + 0x3E (0x00007ff9`d12f59ae)
CORECLR! CoreLibBinder::GetElementType + 0x1A (0x00007ff9`d131f90a)
CORECLR! SigPointer::GetTypeHandleThrowing + 0x937 (0x00007ff9`d1664dc7)
CORECLR! SigPointer::GetTypeHandleThrowing + 0x19B0 (0x00007ff9`d1665e40)
CORECLR! ClassLoader::LoadTypeDefOrRefOrSpecThrowing + 0x9B5 (0x00007ff9`d13fc4c5)
CORECLR! MethodTableBuilder::LoadExactInterfaceMap + 0x5D9 (0x00007ff9`d192d469)
CORECLR! ClassLoader::LoadExactParentAndInterfacesTransitively + 0x654 (0x00007ff9`d13cc344)
CORECLR! ClassLoader::LoadExactParents + 0x431 (0x00007ff9`d13cc7d1)
CORECLR! ClassLoader::DoIncrementalLoad + 0x78D (0x00007ff9`d13ee7fd)
CORECLR! ClassLoader::LoadTypeHandleForTypeKey_Body + 0xDDC (0x00007ff9`d140059c)
    File: C:\Users\tagoo\source\repos\runtime\src\coreclr\vm\binder.h Line: 393
    Image: C:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe
```

I'm unsure what's "special" here that's causing the error. it seems to occur just by adding a generic type parameter to `IShiftOperators`

**Comment by tannergooding** [comment]
Stack trace here is the following with it seemingly occurring just as part of startup:
```
> coreclr.dll!DbgAssertDialog(const char * szFile, int iLine, const char * szExpr) Line 599	C++
  coreclr.dll!CoreLibBinder::GetExistingClass(BinderClassID id) Line 393	C++
  coreclr.dll!CoreLibBinder::GetElementType(CorElementType type) Line 218	C++
  coreclr.dll!SigPointer::GetTypeHandleThrowing(Module * pModule, const SigTypeContext * pTypeContext, ClassLoader::LoadTypesFlag fLoadTypes, ClassLoadLevel level, int dropGenericArgumentLevel, const Substitution * pSubst, const ZapSig::Context * pZapSigContext, MethodTable * pMTInterfaceMapOwner) Line 1091	C++
  coreclr.dll!SigPointer::GetTypeHandleThrowing(Module * pModule, const SigTypeContext * pTypeContext, ClassLoader::LoadTypesFlag fLoadTypes, ClassLoadLevel level, int dropGenericArgumentLevel, const Substitution * pSubst, const ZapSig::Context * pZapSigContext, MethodTable * pMTInterfaceMapOwner) Line 1388	C++
  coreclr.dll!ClassLoader::LoadTypeDefOrRefOrSpecThrowing(Module * pModule, unsigned int typeDefOrRefOrSpec, const SigTypeContext * pTypeContext, ClassLoader::NotFoundAction fNotFoundAction, ClassLoader::PermitUninstantiatedFlag fUninstantiated, ClassLoader::LoadTypesFlag fLoadTypes, ClassLoadLevel level, int dropGenericArgumentLevel, const Substitution * pSubst, MethodTable * pMTInterfaceMapOwner) Line 2035	C++
  coreclr.dll!MethodTableBuilder::LoadExactInterfaceMap(MethodTable * pMT) Line 9208	C++
  coreclr.dll!ClassLoader::LoadExactParentAndInterfacesTransitively(MethodTable * pMT) Line 908	C++
  coreclr.dll!ClassLoader::LoadExactParents(MethodTable * pMT) Line 941	C++
  coreclr.dll!ClassLoader::DoIncrementalLoad(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel currentLevel) Line 2871	C++
  coreclr.dll!ClassLoader::LoadTypeHandleForTypeKey_Body(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel targetLevel) Line 3516	C++
  coreclr.dll!ClassLoader::LoadTypeHandleForTypeKey(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel targetLevel, const InstantiationContext * pInstContext) Line 3235	C++
  coreclr.dll!ClassLoader::LoadTypeDefThrowing(Module * pModule, unsigned int typeDef, ClassLoader::NotFoundAction fNotFoundAction, ClassLoader::PermitUninstantiatedFlag fUninstantiated, unsigned int tokenNotToLoad, ClassLoadLevel level, Instantiation * pTargetInstantiation) Line 2216	C++
  coreclr.dll!SigPointer::GetTypeHandleThrowing(Module * pModule, const SigTypeContext * pTypeContext, ClassLoader::LoadTypesFlag fLoadTypes, ClassLoadLevel level, int dropGenericArgumentLevel, const Substitution * pSubst, const ZapSig::Context * pZapSigContext, MethodTable * pMTInterfaceMapOwner) Line 1422	C++
  coreclr.dll!ClassLoader::LoadTypeDefOrRefOrSpecThrowing(Module * pModule, unsigned int typeDefOrRefOrSpec, const SigTypeContext * pTypeContext, ClassLoader::NotFoundAction fNotFoundAction, ClassLoader::PermitUninstantiatedFlag fUninstantiated, ClassLoader::LoadTypesFlag fLoadTypes, ClassLoadLevel level, int dropGenericArgumentLevel, const Substitution * pSubst, MethodTable * pMTInterfaceMapOwner) Line 2035	C++
  coreclr.dll!MethodTableBuilder::LoadExactInterfaceMap(MethodTable * pMT) Line 9208	C++
  coreclr.dll!ClassLoader::LoadExactParentAndInterfacesTransitively(MethodTable * pMT) Line 908	C++
  coreclr.dll!ClassLoader::LoadExactParents(MethodTable * pMT) Line 941	C++
  coreclr.dll!ClassLoader::DoIncrementalLoad(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel currentLevel) Line 2871	C++
  coreclr.dll!ClassLoader::LoadTypeHandleForTypeKey_Body(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel targetLevel) Line 3516	C++
  coreclr.dll!ClassLoader::LoadTypeHandleForTypeKey(TypeKey * pTypeKey, TypeHandle typeHnd, ClassLoadLevel targetLevel, const InstantiationContext * pInstContext) Line 3235	C++
  coreclr.dll!ClassLoader::EnsureLoaded(TypeHandle typeHnd, ClassLoadLevel level) Line 1021	C++
  coreclr.dll!CoreLibBinder::LoadPrimitiveType(CorElementType et) Line 1194	C++
  coreclr.dll!SystemDomain::LoadBaseSystemClasses() Line 1378	C++
  coreclr.dll!SystemDomain::Init() Line 1203	C++
  coreclr.dll!EEStartupHelper() Line 935	C++
  coreclr.dll!`EEStartup'::`9'::__Body::Run(void * p) Line 1073	C++
  coreclr.dll!EEStartup() Line 1075	C++
  coreclr.dll!EnsureEEStarted() Line 311	C++
  coreclr.dll!CorHost2::Start() Line 101	C++
```

**Comment by tannergooding** [comment]
Should probably tag @trylek as well

**Comment by tannergooding** [comment]
Logged https://github.com/dotnet/runtime/issues/71441

File: src/coreclr/vm/appdomain.cpp [hunk 1]
```diff
@@ -1373,6 +1373,9 @@ void SystemDomain::LoadBaseSystemClasses()
     // We have delayed allocation of CoreLib's static handles until we load the object class
     CoreLibBinder::GetModule()->AllocateRegularStaticHandles(DefaultDomain());
 
+    // Int32 has to be loaded first to break cycle in IShiftOperators
+    CoreLibBinder::LoadPrimitiveType(ELEMENT_TYPE_I4);
```
Comment (dakersnar) [inline src/coreclr/vm/appdomain.cpp:1377]: What's this change?

**Review by dakersnar — APPROVED** [review]
Looks good. All of our types here use `int` as `TOther`, right? Is the intent to allow users implementing the interface to use different types as the `shiftAmount`?

File: src/coreclr/vm/appdomain.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/vm/appdomain.cpp:1377]: A change from Jan that fixes a VM bug that was preventing this change from working.

**Pull request merged**